src/fsharp/FSharp.Compiler.Service.ProjectCracker/Program.fs

```diff
@@ -0,0 +1,414 @@
+﻿namespace FSharp.Compiler.Service.ProjectCracker
+
+open Microsoft.Build.Framework
+open Microsoft.Build.Utilities
+open System.Text
+open System.IO
+open System
+open System.Reflection
+open System.Runtime.Serialization.Formatters.Binary
+open System.Runtime.Serialization.Json
+
+module Program =
+  let runningOnMono = 
+      try match System.Type.GetType("Mono.Runtime") with null -> false | _ -> true
+      with e -> false
+
+  type internal BasicStringLogger() =
+    inherit Logger()
+
+    let sb = new StringBuilder()
+
+    let log (e: BuildEventArgs) =
+      sb.Append(e.Message) |> ignore
+      sb.AppendLine() |> ignore
+
+    override x.Initialize(eventSource:IEventSource) =
+      sb.Clear() |> ignore
+      eventSource.AnyEventRaised.Add(log)
+    
+    member x.Log = sb.ToString()
+
+  type internal HostCompile() =
+      member th.Compile(_, _, _) = 0
+      interface ITaskHost
+
+  //----------------------------------------------------------------------------
+  // FSharpProjectFileInfo
+  //
+  [<Sealed; AutoSerializable(false)>]
+  type FSharpProjectFileInfo (fsprojFileName:string, ?properties, ?enableLogging) =
+
+      let properties = defaultArg properties []
+      let enableLogging = defaultArg enableLogging false
+      let mkAbsolute dir v = 
+          if Path.IsPathRooted v then v
+          else Path.Combine(dir, v)
+
+      let mkAbsoluteOpt dir v =  Option.map (mkAbsolute dir) v
+
+      let logOpt =
+          if enableLogging then
+              let log = new BasicStringLogger()
+              do log.Verbosity <- Microsoft.Build.Framework.LoggerVerbosity.Diagnostic
+              Some log
+          else
+              None
+
+      // Use the old API on Mono, with ToolsVersion = 12.0
+      let CrackProjectUsingOldBuildAPI(fsprojFile:string) = 
+          let engine = new Microsoft.Build.BuildEngine.Engine()
+          try
+             engine.DefaultToolsVersion <- "12.0"
```

Comment:
If its set to 4.0 then the assembly redirects take care of the correct
version loading, it you set it to 12.0 then your moving the default tools
version too high for those tasks which dont explicitly set one.

On 19 November 2015 at 11:39, Robin Neatherway <notifications@github.com>
wrote:

> In src/fsharp/FSharp.Compiler.Service.ProjectCracker/Program.fs
> <https://github.com/fsharp/FSharp.Compiler.Service/pull/464#discussion_r45329927>
> :
>
> > +
> > +      let mkAbsoluteOpt dir v =  Option.map (mkAbsolute dir) v
> > +
> > +      let logOpt =
> > +          if enableLogging then
> > +              let log = new BasicStringLogger()
> > +              do log.Verbosity <- Microsoft.Build.Framework.LoggerVerbosity.Diagnostic
> > +              Some log
> > +          else
> > +              None
> > +
> > +      // Use the old API on Mono, with ToolsVersion = 12.0
> > +      let CrackProjectUsingOldBuildAPI(fsprojFile:string) =
> > +          let engine = new Microsoft.Build.BuildEngine.Engine()
> > +          try
> > +             engine.DefaultToolsVersion <- "12.0"
>
> Setting this to 12.0 here means that version 4.0 msbuild tasks will throw
> a invalid cast exception
>
> Ah, I thought the assembly binding redirect would address that?
>
> I originally set it to 12.0 because otherwise mono msbuild didn't work
> correctly on projects with ToolsVersion set to 12.0 (hence the test with
> the same name in this repo). Is this fixed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fsharp/FSharp.Compiler.Service/pull/464/files#r45329927>
> .
>